### ISSUE TITLE
[Autofill] Switch to `SecSharedCredentials` on iOS, and explore how we can expose it in the framework

### ISSUE THREAD

**Issue by BelfDev** [flutter/flutter#72364]
## Use case

I work on a multi-platform e-commerce product that is available in several countries and has many associated domains. When using `AutofillGroup` as recommended by the [docs](https://api.flutter.dev/flutter/widgets/AutofillGroup-class.html) to save login/signup credentials for the first time, they get saved in association with the wrong domain. 

This is problematic because it can cause confusion to customers. To illustrate, consider the example below.

- Users from Germany access `company.de` on Safari and register their credentials using the keychain (which correctly associates these credentials with the `.de` domain).
- These users decide to access the product via the Flutter mobile App. When they try to log in on iOS, the keychain correctly identifies that the App has the `.de` associated domain and suggests the previously registered credentials.
- Users accept the suggestion and get their username/password auto-filled. 
- However, when pressing on the submission button, they are asked if they want to save new credentials because the **keychain is trying to store them in the wrong domain** (e.g. `company.ie`). 

_"If I have already saved my credentials for that product/service and nothing has changed, why am I being asked to save them again?"_

## Proposal

We need a way to **set which associated domain we want to use for storing credentials** with password managers and `AutofillGroup`

If a new user accesses the App with the `de_DE` locale, his/her credentials must be saved to the appropriate `company.de` domain instead of the other available options.

**Comment by HansMuller** [comment]
CC @LongCatIsLooong

**Comment by lunaticcoding** [comment]
I have a similar issue and would love this feature :D please :)

**Comment by LongCatIsLooong** [comment]
I looked at https://developer.apple.com/documentation/safariservices/supporting_associated_domains and https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_developer_associated-domains, there currently doesn't seem to be a way to "bind" an associated domain to a locale on iOS? 

@BelfDev do you happen to know if this is doable in an iOS `UITextField`?

**Comment by BelfDev** [comment]
@LongCatIsLooong indeed, there isn't a standard way of binding an associated domain to a locale on iOS because the domain format might vary considerably -- making it virtually impossible for the platform to infer the correct binding. However, we might be able to delegate a custom binding mechanism to the user by combining two characteristics of Apple's API:

> iOS supports Password AutoFill on UITextField, UITextView, and any custom view that adopts the UITextInput protocol. ([source](https://developer.apple.com/documentation/security/password_autofill/enabling_password_autofill_on_a_text_input_view)).

> The **Security.SecSharedCredentials API** provides functions for storing and requesting shared password-based credentials. ([source](https://developer.apple.com/documentation/security/shared_web_credentials?language=objc))

So, in theory, we could include in the custom `UITextInput` implementation a method that handled the binding. This method could use [SecAddSharedWebCredential](https://developer.apple.com/documentation/security/1617986-secaddsharedwebcredential?language=objc) to manually save credentials to the correct associated domain. 

From the code-user perspective on Flutter, one could provide an `AutofillDelegate` to indicate how to link the `locale` to a supported `associated-domain`.

E.g. `de_DE` → `company.de`

### Useful resources
- Apple App site association validator tool. Click [here](https://branch.io/resources/aasa-validator/).
- Apple search validator tool. Click [here](https://search.developer.apple.com/appsearch-validation-tool).
- An example code snippet with `SecRequestSharedWebCredential` use. Click [here](https://github.com/mluedke2/ultra-motivator/blob/e8a7f845c504f64a46fe67433994fc63c92f2c01/ios/UltraMotivator/UltraMotivator/SafariKeychainManager.swift#L13-L38).

**Comment by LongCatIsLooong** [comment]
Repurposing this issue to:
- Change the iOS autofill save implementation to use `SecSharedCredentials` 
- Explore how we can expose the API in the framework (since it's platform specific), or make the platform side extensible.
